### PR TITLE
add `evaluate` for `FreeAssAlgElem`

### PR DESCRIPTION
**Pull request by ThomasBreuer**

### CODE REVIEW

**Comment by fingolfin** [comment]
Tests fail :-(

**Comment by ThomasBreuer** [comment]
> Tests fail :-(

Yes.
I do not really agree with Aqua.jl here, but I will adjust the code to it.

**Comment by fingolfin** [comment]
If there is a bug in Aqua, we should investigate it (CC @lgoettgens).

Right now it reports:
```
Unbound type parameters detected:
[1] (a::FreeAssAlgElem{T})(vals::U...) where {T<:RingElement, U<:NCRingElem} in AbstractAlgebra at /home/runner/work/AbstractAlgebra.jl/AbstractAlgebra.jl/src/FreeAssAlgebra.jl:208
Unbound type parameters: Test Failed at /home/runner/.julia/packages/Aqua/eC6PS/src/unbound_args.jl:50
  Expression: isempty(unbounds)
   Evaluated: isempty(Method[(a::FreeAssAlgElem{T})(vals::U...) where {T<:RingElement, U<:NCRingElem} in AbstractAlgebra at /home/runner/work/AbstractAlgebra.jl/AbstractAlgebra.jl/src/FreeAssAlgebra.jl:208])
Stacktrace:
 [1] test_unbound_args(m::Module; broken::Bool)
   @ Aqua ~/.julia/packages/Aqua/eC6PS/src/unbound_args.jl:50
 [2] test_unbound_args
   @ ~/.julia/packages/Aqua/eC6PS/src/unbound_args.jl:36 [inlined]
 [3] macro expansion
   @ ~/.julia/packages/Aqua/eC6PS/src/Aqua.jl:73 [inlined]
 [4] macro expansion
   @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [5] test_all(testtarget::Module; ambiguities::Bool, unbound_args::Bool, undefined_exports::Bool, project_extras::Bool, stale_deps::Bool, deps_compat::Bool, project_toml_formatting::Bool, piracy::Bool)
   @ Aqua ~/.julia/packages/Aqua/eC6PS/src/Aqua.jl:72
```

The function it complains about:
```julia
function (a::AbstractAlgebra.FreeAssAlgElem{T})(vals::U...) where {T <: RingElement, U <: NCRingElem}
   return evaluate(a, [vals...])
end
```
I don't understand why it complains...? Admittedly I would also drop the `T` restriction (to make the code more general), but that doesn't mean it should be forbidden to restrict it...?

**Comment by lgoettgens** [comment]
> If there is a bug in Aqua, we should investigate it (CC @lgoettgens).
> 
> Right now it reports:
> 
> ```
> Unbound type parameters detected:
> [1] (a::FreeAssAlgElem{T})(vals::U...) where {T<:RingElement, U<:NCRingElem} in AbstractAlgebra at /home/runner/work/AbstractAlgebra.jl/AbstractAlgebra.jl/src/FreeAssAlgebra.jl:208
> Unbound type parameters: Test Failed at /home/runner/.julia/packages/Aqua/eC6PS/src/unbound_args.jl:50
>   Expression: isempty(unbounds)
>    Evaluated: isempty(Method[(a::FreeAssAlgElem{T})(vals::U...) where {T<:RingElement, U<:NCRingElem} in AbstractAlgebra at /home/runner/work/AbstractAlgebra.jl/AbstractAlgebra.jl/src/FreeAssAlgebra.jl:208])
> Stacktrace:
>  [1] test_unbound_args(m::Module; broken::Bool)
>    @ Aqua ~/.julia/packages/Aqua/eC6PS/src/unbound_args.jl:50
>  [2] test_unbound_args
>    @ ~/.julia/packages/Aqua/eC6PS/src/unbound_args.jl:36 [inlined]
>  [3] macro expansion
>    @ ~/.julia/packages/Aqua/eC6PS/src/Aqua.jl:73 [inlined]
>  [4] macro expansion
>    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
>  [5] test_all(testtarget::Module; ambiguities::Bool, unbound_args::Bool, undefined_exports::Bool, project_extras::Bool, stale_deps::Bool, deps_compat::Bool, project_toml_formatting::Bool, piracy::Bool)
>    @ Aqua ~/.julia/packages/Aqua/eC6PS/src/Aqua.jl:72
> ```
> 
> The function it complains about:
> 
> ```julia
> function (a::AbstractAlgebra.FreeAssAlgElem{T})(vals::U...) where {T <: RingElement, U <: NCRingElem}
>    return evaluate(a, [vals...])
> end
> ```
> 
> I don't understand why it complains...? Admittedly I would also drop the `T` restriction (to make the code more general), but that doesn't mean it should be forbidden to restrict it...?

It complains about U not being present in the 0-parameter vararg case. One should split up the function into:
`function(::FreeAssAlgElem{T})() where {T <: RingElement}` and function(::FreeAssAlgElem{T}(val::U, vals::U...) where {...}`.
The second one states that it has at least one parameter in the vararg.
I have done similar things at other places, the first I can think of is https://github.com/oscar-system/Oscar.jl/blob/8f1e31acb687d8efd002402cda2277c6b430199a/experimental/LieAlgebras/src/LieAlgebraModule.jl#L801-L804. The ones Aqua complains about because of unbound type parameters in all our packages are on my todo list.

**Comment by lgoettgens** [comment]
This is even stated in the Aqua docs: https://juliatesting.github.io/Aqua.jl/dev/#Aqua.test_unbound_args-Tuple{Module}

**Comment by ThomasBreuer** [comment]
Yes, I know Aqua's arguments, but as I said, I do not agree with them.
There would be a problem with the prescribed common type `U` of all arguments if `U` would be used inside the method, because there is no way to determine `U` in the case of zero arguments.
However, I think there is no problem if `U` is not used inside the method but in order to express that all argumnts should have the same type.

**Comment by lgoettgens** [comment]
I have read somewhere (unfortunately no idea when and where), that even in your use case, the unbound type parameter can result in degraded performance. No idea, how much and how relevant to fix.

**Comment by fingolfin** [comment]
Ahh, of course, I keep forgetting about this, thanks for me reminding me, @lgoettgens.

In this case I do think there is a genuine problem in so far as in the case of zero arguments of type `U`, we cannot (in general) determine in which ring the output should sit, and cannot do anything meaningful. So, to solve this, we should just change the method signature to `function(::FreeAssAlgElem{T}(val::U, vals::U...) where {...}`.

File: src/FreeAssAlgebra.jl [hunk 1]
```diff
@@ -149,6 +149,66 @@ function Base.hash(x::FreeAssAlgElem{T}, h::UInt) where T <: RingElement
...
+   r = zero(S)
+   o = one(S)
+   for (c, v) in zip(coefficients(a), exponent_words(a))
+      r = addeq!(r, c*prod((vals[i] for i in v), init = o))
+   end
+   return r
+end
+
+function (a::AbstractAlgebra.FreeAssAlgElem{T})(vals::U...) where {T <: RingElement, U <: NCRingElem}
+   return evaluate(a, [vals...])
```
Comment (fingolfin) [inline src/FreeAssAlgebra.jl:209]:
```suggestion
function (a::AbstractAlgebra.FreeAssAlgElem{T})(val::U, vals::U...) where {T <: RingElement, U <: NCRingElem}
   return evaluate(a, [val, vals...])
```

**Comment by thofma** [comment]
Looks good to me, but I agree that the 0-arg version needs to be adjusted as suggested by @fingolfin (I know it is annoying).

**Comment by ThomasBreuer** [comment]
Just a conceptual comment:
Since `free_associative_algebra(QQ, Symbol[])` is a valid input, also evaluating elements of such algebras makes sense.
(`polynomial_ring` requires at least one variable.)
As is mentioned above, the syntax `evaluate(elm, vars, vals)` is not suitable for that because an empty `vals` does not provide enough  information.
GAP supports `Value( elm, vars, vals[, one] )` where the optional argument `one` means the intended image of the identity.
AbstractAlgebra's `_evaluate` seems to follow the same idea, but this is used only in special cases.

**Pull request merged**